### PR TITLE
Granular Auth

### PR DESCRIPTION
**Pull request by iamjohnnym**
added connections object
added ability to use sts or not by designation of a flag, `self.args['assume']`

### CODE REVIEW

**Comment by theherk** [comment]
This is not my library, but I got notified as a contributor, and I have a few suggestions:

Some of the formatting is not PEP-8 compliant.
- function names should be `does_foo()` rather than `doesFoo`
- line continuation is a bit odd in places

I don't understand why set a variable in this case:

``` python
    self.client = boto3.client(self.args['aws_asset'], region_name=self.args['region'])
        return self.client
```

rather than just:

``` python
    return boto3.client(self.args['aws_asset'], region_name=self.args['region'])
```

Setting `self.args = kwargs` conflates the two separate meanings in Python. It is clearer to just set a variable for each value, I believe.

Should probably raise an exception rather than print when something goes wrong so the program importing this library can handle.

connections.py has no entry point so there is no need for the hashbang.

Also, I'm not sure what is gained by teaching this library how to use STS. There is already a great library for that, boto3. With it you can assume a role then pass those credentials into this object at init time. It may be a great idea; I just don't fully understand.

Also, all of you commits are done by not you Github email address, so they aren't properly linked to your user. You can use this script to correct that: [Changing author info](https://help.github.com/articles/changing-author-info/).

File: awsrequests/connections.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                aws_access_key_id=self.temp_creds["Credentials"]["AccessKeyId"],
+                aws_secret_access_key=self.temp_creds["Credentials"]
+                                                     ['SecretAccessKey'],
+                aws_session_token=self.temp_creds["Credentials"]["SessionToken"],
+                region_name=self.args['region']
+                )
+        return self.client
+
+    def login(self):
+        if not self.args.has_key('assume') \
```
Comment (djcrabhat) [inline awsrequests/connections.py:46]: I think it's a little more pythonic to do it this way:

```
if "assume" in args and args["assume"]:
    # do stuff for assuming
    pass
else:
    # do the other stuff
    pass
```

File: awsrequests/__init__.py [hunk 2]
```diff
@@ -122,15 +115,18 @@ def request(self, method, url,
...
-                                                   self.secret_key,
-                                                   self.session_token,
-                                                   payload=prepped.body,
-                                                   method=prepped.method,
-                                                   t=time)
+        aws_auth_headers = get_headers_for_request(
+                url,
+                self.args['region'],
+                'execute-api',
+                self.temp_creds["Credentials"]["AccessKeyId"],
```
Comment (djcrabhat) [inline awsrequests/__init__.py:122]: You can't force using temp creds here.  This library used to just piggy-back off boto, so you could use real creds or meta-data provided creds or whatever.  This forces to uses STS, which is not right

File: awsrequests/connections.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env python
+
+from botocore.exceptions import ClientError
+import boto3
+
+class AWSConnections(object):
+    def __init__(self, **kwargs):
+        self.args = kwargs
+        if 'client' in self.args['resource']:
+            self.client = self.setClient()
```
Comment (djcrabhat) [inline awsrequests/connections.py:10]: Another thing I don't understand is what this client is used for.  We don't make any AWS API calls in this library beyond the sts-ing, and I don't see it ever used anywhere, only ever assigned.

File: awsrequests/connections.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+                aws_secret_access_key=self.temp_creds["Credentials"]
+                                                     ['SecretAccessKey'],
+                aws_session_token=self.temp_creds["Credentials"]["SessionToken"],
+                region_name=self.args['region']
+                )
+        return self.client
+
+    def login(self):
+        if not self.args.has_key('assume') \
+           or self.args['assume'] is False:
```
Comment (djcrabhat) [inline awsrequests/connections.py:47]: And moreover, I don't think you need this extra param to tell it to assume.  Just check for the existence of your "role_arn" param being set.  If it doesn't exist, you're not trying to assume.  If it does exist, you're trying to assume to that role.

File: awsrequests/__init__.py [hunk 5]
```diff
@@ -22,48 +23,40 @@ def emit(self, record):
...
-            self.session_expires = temp_security_creds["Credentials"]["Expiration"]
-        except botocore.exceptions.BotoCoreError as e:
-            raise Exception("error assuming role: {0}".format(e.args))
+class AwsRequester(AWSConnections):
+    def __init__(self, **kwargs):
+        """
+        **kwargs gets set to self.args as a dict
+
+        REQUIRED
+        :param: aws_asset = 'ec2', 'sts', 's3', 'dynamodb', any of the services
```
Comment (djcrabhat) [inline awsrequests/__init__.py:32]: So this library _can_ call to AWS's own APIs, but that's really not it's sweetspot.  Boto is so much better at that.  This is really just to give a wrapper for the requests (http://docs.python-requests.org/en/master/) library, for the primary purpose of talking to your own API Gateways that are protected by IAM.  This `aws_asset` is only ever used for construction of that client, which doesn't seem to get used anywhere.

File: awsrequests/__init__.py [hunk 6]
```diff
@@ -22,48 +23,40 @@ def emit(self, record):
...
+                            that amazon offers through boto3
+        :param: region = region to connect to
+        :param: aws_account = aws account # to access
+        :param: role = role used for your arn
```
Comment (djcrabhat) [inline awsrequests/__init__.py:36]: I don't know what "role used for your arn" means

File: awsrequests/__init__.py [hunk 7]
```diff
@@ -22,48 +23,40 @@ def emit(self, record):
...
+        :param: role_name = the name in which will be seen through the
```
Comment (djcrabhat) [inline awsrequests/__init__.py:37]: this is standardly known as the "Role Session Name" in AWS-land

**Comment by djcrabhat** [comment]
I think this PR misses the point of the original library.  The library as currently constituted piggy backs off of the process described in the boto3 docs (http://boto3.readthedocs.io/en/latest/guide/configuration.html) to resolve credentials.  I had the assume_role function there before to do help facilitate those kinda "double hop" role assumptions.  Most of the time, you'd just have an environment variable set or getting creds from metadata service, so you'd never even think about it.    

Also, it goes from taking explicit, plain params in the construction of the  AwsRequests object to a kwargs that really makes it much more awkward to use.  And the params of that kwargs just seem to be incongruous to the intention of this lib.

**Comment by djcrabhat** [comment]
Gonna close as I don't think this is gonna go forward like this.  Hit me up with a new PR if you're really excited about doing it, and we can reconsider.

**Pull request closed without merging**